client/src/client.rs

```diff
@@ -197,11 +204,296 @@ impl<T: StorageType> Client<T> {
     pub async fn send_message(
         &mut self,
         message: &str,
-        user_id: &str,
-        device_id: u32,
+        service_id: &ServiceId,
     ) -> Result<(), SignalClientError> {
-        self.server_api
-            .send_msg(message.into(), user_id.into(), device_id)
+        let content = Content::builder()
+            .data_message(
+                DataMessage::builder()
+                    .body(message.to_owned())
+                    .contact(vec![])
+                    .body_ranges(vec![])
+                    .preview(vec![])
+                    .attachments(vec![])
+                    .build(),
+            )
+            .build();
+
+        let timestamp = SystemTime::now();
+
+        // Update the contact.
+        let to = match self.contact_manager.get_contact(service_id) {
+            Err(_) => {
+                self.contact_manager
+                    .add_contact(service_id, 1.into())
+                    .expect("Can add contact that does not exist yet");
+                let bundles = self
+                    .server_api
+                    .fetch_pre_key_bundles(service_id)
+                    .await
+                    .unwrap();
+
+                let mut device_ids = Vec::new();
+                for ref bundle in bundles {
+                    device_ids.push(bundle.device_id().unwrap());
+                    process_prekey_bundle(
+                        &ProtocolAddress::new(
+                            service_id.service_id_string(),
+                            bundle.device_id().unwrap(),
+                        ),
+                        &mut self.storage.protocol_store.session_store,
+                        &mut self.storage.protocol_store.identity_key_store,
+                        bundle,
+                        SystemTime::now(),
```

Comment:
Maybe save this in a variable outside the loop such that it shows the keys as being created at the same point in time.

Reply:
Maybe or should?